packages/flutter/lib/src/services/text_editing_value.dart

```diff
@@ -0,0 +1,444 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+import 'dart:ui' show
+  TextAffinity,
+  hashValues;
+
+import 'package:characters/characters.dart';
+import 'package:flutter/foundation.dart';
+
+import 'text_editing.dart';
+import 'text_metrics.dart';
+
+TextAffinity? _toTextAffinity(String? affinity) {
+  switch (affinity) {
+    case 'TextAffinity.downstream':
+      return TextAffinity.downstream;
+    case 'TextAffinity.upstream':
+      return TextAffinity.upstream;
+  }
+  return null;
+}
+/// The current text, selection, and composing state for editing a run of text.
+@immutable
+class TextEditingValue {
```

Comment:
Maybe we should have a `TextEditingValue replace(TextRange, String)` method like Android does? One of the biggest pain points from my personal experience is input manipulation is tricky when you have to take the composing range and selection range into account when you just want to change the text.

Reply:
I'd like to hold off on this to avoid increasing the size of this PR any more if that's ok with you, but I really like the idea.  Lots of the existing code in the delete methods here spend a lot of lines getting the composing region right.  It would be cool to just remove the relevant text from the string and then call `replace` and let it figure out composing and selection.  I'll keep an eye out for other opportunities for this.

Reply:
Yeah the android `Editable` implementation implements most common text manipulation methods in terms of `replace`, since it's `replace` implementation takes care of the selection and composition.

Reply:
I investigated adding `replace` to this PR again, but it is making this PR even bigger and more confusing.  I'll follow up in another PR based on my branch https://github.com/justinmc/flutter/tree/text-editing-value-replace.  In the meantime I've removed all changes to TextEditingValue.  `deleteTo` and related stuff is not private inside of TextEditingActionTarget.  In the new PR, I'll replace that with calls to `replace`.